dqn.py

```diff
@@ -0,0 +1,202 @@
+import torch
+import random
+import torch.nn as nn
+import torch.optim as optim
+import torch.nn.functional as F
+import numpy as np
+import re
+import os
+
+
+# DQN with tanh activation function
+def get_activ(a):
+    if a == 't':
+        return nn.Tanh(), 'nn.Tanh()'
+    elif a == 'r':
+        return nn.ReLU(), 'nn.ReLU()'
+    elif a == 'l':
+        return nn.LeakyReLU(), 'nn.LeakyReLU()'
+    elif a == 'e':
+        return nn.ELU(), 'nn.ELU()'
+    elif a == 's':
+        return nn.Sigmoid(), 'nn.Sigmoid()'
+    else:
+        raise ValueError('Unknown activation function: ' + a)
+
+
+def parse_string(hs: str):
+    # split hs into a list of ints and letters
+    h = re.findall(r'\d+|\D+', hs)
+    # check all odd id are letter
+    assert all([h[i].isalpha() for i in range(0, len(h), 2)])
+    # check all even id are int
+    assert all([h[i].isdigit() for i in range(1, len(h), 2)])
+    activ = [h[i] for i in range(0, len(h), 2)]
+    sizes = [int(h[i]) for i in range(1, len(h), 2)]
+    assert len(activ) == len(sizes)
+    return activ, sizes
+
+
+class DQN(nn.Module):
+
+    def __init__(self, input_dim, hs: str, output_dim):
```

Comment:
Your init function for your network is somewhat confusing. It would be nice to see some more descriptive comments here.

Reply:
Hi Dennis, thank you so much for your detailed review. This init function works as this, given a hs string, for example, "t64t64", it will be devides into two list: ['t', 't'], ['64, '64], then it will firstly add an input layer, and then iterate these two lists, to add the first 't' (tanh), then the first '64' (Linear 64), and until finishing two lists, and eventually add the output layer.

Reply:
Oh thats cool I've never seen that way to do it. Takes up less space. Nice job.